### PR TITLE
Handle Player Death

### PR DESCRIPTION
**Pull request by arejayelle**
### NOTE

You cannot load a file back if you died 
Try again works fine
 
### Description
Handle death
Change the "first floor" in the scene manager first floor Choose FloorA B or C
image

### CODE REVIEW

File: Hush/Assets/Scripts/Game/Models/GameState.cs [hunk 1]
```diff
@@ -112,5 +121,15 @@ public GameState()
             randomSceneSeed = 0;
             sceneProgress = -1;
         }
+        private void UnsetSpells()
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/Models/GameState.cs:124]:
```suggestion

        private void UnsetSpells()
```

File: Hush/Assets/Scripts/Game/Partials/GameManager.Scene.cs [hunk 2]
```diff
@@ -32,7 +33,7 @@ public void ResetSceneProgress()
 
         private void ApplySceneState()
         {
-            SceneManager.Instance.LoadScene(SaveGameManager.Instance.SavedGameScene);
+            SceneManager.Instance.LoadScene( SaveGameManager.Instance.SavedGameScene);
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/Partials/GameManager.Scene.cs:36]:
```suggestion
            SceneManager.Instance.LoadScene(SaveGameManager.Instance.SavedGameScene);
```

File: Hush/Assets/Scripts/Game/Partials/GameManager.Transitions.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        /// </summary>
+        public void OnRunCompleted()
+        {
+            Instance.ResetTemporaryBuffs();
+            Instance.RestorePlayerHealth();
+            Instance.ResetSceneProgress();
+            SaveGameManager.Instance.OnSave();
+        }
+
+        private void LoadGameOver()
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/Partials/GameManager.Transitions.cs:16]: Should be in SceneManager has nothing to do with the state really

File: Hush/Assets/Scripts/Game/SceneManager.cs [hunk 4]
```diff
@@ -7,14 +8,14 @@ public class SceneManager : MonoBehaviour
...
-        [Header("Scene Build Indexes")]
-        [SerializeField] private int[] levelSceneIndexes;
-        [SerializeField] private int mainMenuIndex;
-        [SerializeField] private int finalSceneIndex;
-        [SerializeField] private int endgameSceneIndex;
-        
-        [Header("New Game Parameters")]
-        [SerializeField] private SpellType initialLightSpell = SpellType.FireballSpell;
+        [Header("Scene Build Indexes")] [SerializeField]
+        private Scenes[] levelSceneIndexes;
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/SceneManager.cs:12]:
```suggestion
        [Header("Scene Build Indexes")] 
        [SerializeField] private Scenes[] levelSceneIndexes;
```

File: Hush/Assets/Scripts/Game/SceneManager.cs [hunk 5]
```diff
@@ -7,14 +8,14 @@ public class SceneManager : MonoBehaviour
...
+
+        [Header("New Game Parameters")] [SerializeField]
+        private SpellType initialLightSpell = SpellType.FireballSpell;
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/SceneManager.cs:15]:
```suggestion
        [Header("New Game Parameters")] 
        [SerializeField] private SpellType initialLightSpell = SpellType.FireballSpell;
```

File: Hush/Assets/Scripts/Game/SceneManager.cs [hunk 6]
```diff
@@ -7,14 +8,14 @@ public class SceneManager : MonoBehaviour
...
+
+        [SerializeField] private Scenes firstFloor = Scenes.FloorA;
+
         [SerializeField] private SpellType initialHeavySpell = SpellType.StunSpell;
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/SceneManager.cs:19]:
```suggestion
        [Header("New Game Parameters")] 
        [SerializeField]private SpellType initialLightSpell = SpellType.FireballSpell;
        [SerializeField] private SpellType initialHeavySpell = SpellType.StunSpell;
        [SerializeField] private Scenes firstFloor = Scenes.FloorA;
```

File: Hush/Assets/Scripts/Game/SceneManager.cs [hunk 7]
```diff
@@ -7,14 +8,14 @@ public class SceneManager : MonoBehaviour
...
-        
-        [Header("New Game Parameters")]
-        [SerializeField] private SpellType initialLightSpell = SpellType.FireballSpell;
+        [Header("Scene Build Indexes")] [SerializeField]
+        private Scenes[] levelSceneIndexes;
+
+        [Header("New Game Parameters")] [SerializeField]
+        private SpellType initialLightSpell = SpellType.FireballSpell;
+
+        [SerializeField] private Scenes firstFloor = Scenes.FloorA;
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/SceneManager.cs:17]: Assument the first floor is simply the first entry in levelSceneIndexes

File: Hush/Assets/Scripts/Game/SceneManager.cs [hunk 8]
```diff
@@ -89,46 +92,65 @@ private void LoadFinal()
...
+#endif
         }
 
         private void StartNewGame()
         {
             // TODO Generate seed & randomize scene order based on seed
+
+            // TEMPORARY VALUE REMOVE
+            LoadFirstFloor();
+            GameManager.Instance.SetLoadedScene( Instance.firstFloor);
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/SceneManager.cs]: Remove this before approval

File: Hush/Assets/Scripts/Game/SceneManager.cs [hunk 9]
```diff
@@ -89,46 +92,65 @@ private void LoadFinal()
...
-        private void TransitionToScene(int index)
+        private void TransitionToScene(Scenes sceneIndex)
         {
             // TODO Transition between scenes
-            
-            UnityEngine.SceneManagement.SceneManager.LoadScene(index);
-            
-            GameManager.Instance.SetLoadedScene(index);
-            
+            UnityEngine.SceneManagement.SceneManager.LoadScene( (int) sceneIndex);
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/SceneManager.cs:152]:
```suggestion
            UnityEngine.SceneManagement.SceneManager.LoadScene((int) sceneIndex);
```

File: Hush/Assets/Scripts/Player/PlayerHitPoint.cs [hunk 10]
```diff
@@ -24,7 +24,7 @@ public class PlayerHitPoint : MonoBehaviour, IKillable
 
         #region Public Variables
 
-        private UnityEvent _killed;
+        [SerializeField] private UnityEvent _killed;
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Player/PlayerHitPoint.cs:27]: undo this, access Killed instead of _killed

File: Hush/Assets/Scripts/Doors/Door.cs [hunk 11]
```diff
@@ -83,7 +83,7 @@ private void OpenDoor(InputAction.CallbackContext context)
 
         private void LoadNext()
         {
-            SceneManager.Instance.LoadNextScene();
+            SceneManager.Instance.LoadTempPowerScene();
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Doors/Door.cs:86]: No need to load Power Scene, powers will be in the Scene UI already. Talk to @daniela-venuta

File: Hush/Assets/Scripts/Game/Enums/Scenes.cs [hunk 12]
```diff
@@ -0,0 +1,17 @@
...
+       Win = 1,
+       GameOver = 2,
+       SpellChoice= 3,
+       PowerUp = 4,
+       PermanentPower = 5, // TODO: actually replace it in buildSettings
+       FinalFloor = 6,
+       FloorA = 7,
+       FloorB = 8,
+       FloorC = 9,
+
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/Enums/Scenes.cs:15]:
```suggestion
```

File: Hush/Assets/Scripts/Game/Enums/Scenes.cs [hunk 13]
```diff
@@ -0,0 +1,17 @@
+﻿namespace Game.Enums
+{
+    public enum Scenes
+    {
+       MainMenu = 0,
+       Win = 1,
+       GameOver = 2,
+       SpellChoice= 3,
+       PowerUp = 4,
+       PermanentPower = 5, // TODO: actually replace it in buildSettings
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/Enums/Scenes.cs:10]: No need for this scene

File: Hush/Assets/Scripts/Game/GameManager.cs [hunk 14]
```diff
@@ -49,5 +49,12 @@ public void SetGameState(GameState gameState)
             ApplyEcholocationState();
             ApplySpellState();
         }
+
+        public void NewGame()
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/GameManager.cs:53]: NewGame() should be in SceneManager and should call something like GameManager.Instance.ResetGameState()

which does what you do at line 55

File: Hush/Assets/Scripts/Game/Models/GameState.cs [hunk 15]
```diff
@@ -72,27 +73,35 @@ public GameState()
...
+        private void UnsetBoosts()
+        {
             damageBoost = 0f;
             speedBoost = 0f;
             vitalityBoost = 0f;
+        }
+        private void UnsetFloorStats()
+        {
+            playerTransform = new SerializableTransform(Vector3.zero, Quaternion.identity);
+            isPlayerInvisible = false;
```
Comment (po-trottier) [inline Hush/Assets/Scripts/Game/Models/GameState.cs:100]:
```suggestion

        private void UnsetBoosts()
        {
            damageBoost = 0f;
            speedBoost = 0f;
            vitalityBoost = 0f;
        }
        
        private void UnsetFloorStats()
        {
            playerTransform = new SerializableTransform(Vector3.zero, Quaternion.identity);
            isPlayerInvisible = false;
```

File: Hush/Assets/Scripts/Game/SceneManager.cs (hunk 7 above)
Comment (arejayelle) [reply Hush/Assets/Scripts/Game/SceneManager.cs:17]: Well, it's the first floor you want to load in. But for testing purposes, I was setting the "default first floor" to one of the other ones to make sure they were okay

Comment (po-trottier) [reply Hush/Assets/Scripts/Game/SceneManager.cs:17]: So this is just a testing value

**Review by po-trottier — APPROVED** [review]

**Pull request merged**